metrics/sinks/gcm/gcm.go

```diff
@@ -321,21 +324,37 @@ func CreateGCMSink(uri *url.URL) (core.DataSink, error) {
 		return nil, fmt.Errorf("invalid metrics parameter: %s", metrics)
 	}
 
-	if err := gce_util.EnsureOnGCE(); err != nil {
-		return nil, err
+	var gcpConfig struct {
+		ProjectId string `json:"project_id"`
 	}
 
-	// Detect project ID
-	projectId, err := gce.ProjectID()
+	client, err := google.DefaultClient(oauth2.NoContext, gcm.MonitoringScope)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error creating oauth2 client: %s", err)
 	}
 
-	// Create Google Cloud Monitoring service.
-	client := oauth2.NewClient(oauth2.NoContext, google.ComputeTokenSource(""))
 	gcmService, err := gcm.New(client)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error creating GCM service: %s", err)
+	}
+
+	// Try and get the project ID, GCE metadata first
+	projectId, err := gce.ProjectID()
+	// If there was a problem, try the default credentials file
+	if err != nil {
+		file := os.Getenv(gcpCredentialEnv)
+		if file == "" {
+			return nil, fmt.Errorf("no %s environment variable detected", gcpCredentialEnv)
+		}
+		conf, err := ioutil.ReadFile(file)
+		if err != nil {
+			return nil, fmt.Errorf("error reading default credentials file: %s", err)
+		}
+		err = json.Unmarshal(conf, &gcpConfig)
```

Comment:
Note: older credentials files may not have an explicit project id field, but have it embedded in the email address. Not sure it it's worth handling those cases here.